pages/tokens.py

```diff
@@ -1,169 +1,241 @@
+                       outline=True,
+                       id='token-export', n_clicks=0),
+            dcc.Download(id='download-tokens'),
+        ],
+            style={'display': 'flex', 'gap': '5px', 'margin-bottom': '20px'}
+        ),
+        html.Div(id='token-table'),
+        dbc.Modal([
+            dbc.ModalHeader("Generate tokens"),
+            dbc.ModalBody([
+                html.Div([
+                    dbc.Label('Program'),
+                    dbc.Input(value=(STUDY_CONFIG or 'program'),
+                              id='token-program',
+                              type='text',
+                              disabled=(bool(STUDY_CONFIG)),
+                              required=True),
+                ]),
+                html.Div([
+                    dbc.Label('Token Length'),
+                    dbc.Input(value=10, id='token-length', type='number', min=6, max=100, required=True),
+                ]),
+                html.Div([
+                    dbc.Label('Token Subgroup'),
+                    dcc.Dropdown(
+                        configured_subgroups or ['test'],
+                        id='token-subgroup',
                     ),
+                ]),
+                html.Div([
+                    dbc.Label('Number of Tokens'),
+                    dbc.Input(value=1, id='token-count', type='number', min=0, required=True),
+                ]),
+                dbc.Alert(
+                    id='generate-tokens-alert',
+                    style={'margin': '15px 5px'},
+                ),
+                html.Div([
+                    dbc.Button('Close', id='close-modal-btn', color='primary', outline=True, n_clicks=0),
+                    dbc.Button('Generate', id='generate-tokens-btn', color='primary', n_clicks=0),
+                ],
+                    style={'display': 'flex', 'gap': '10px', 'margin-top': '20px'}
+                ),
+            ],
+                style={'display': 'flex', 'flex-direction': 'column', 'gap': '10px'}
+            )
+        ],
+            id="generate-tokens-modal",
+            is_open=False,
+        )
+    ]
+)
 
-                    html.Br(),
-                    html.Div([
-                        html.Button(children='Generate Tokens', id='token-generate', n_clicks=0, style={
-                            'font-size': '14px', 'width': '140px', 'display': 'block', 'margin-bottom': '10px',
-                            'margin-right': '5px', 'height':'40px', 'verticalAlign': 'top', 'background-color': 'green',
-                            'color': 'white',
-                        }),
-                        dcc.Download(id='download-token'),
-                    ])
 
-                ],
-                xl=3,
-                lg=4,
-                sm=6,
-            ),
-        ]),
+@callback(
+    Output('store-tokens', 'data'),
+    Input('store-tokens', 'data'),
+)
+def load_tokens(_):
+  return [e.get('token') for e in edb.get_token_db().find({})]
```

Comment:
This could be changed to
`return esdt.get_all_tokens()`
so that it is consistent with below